[3D] 3daxis cleanup

ptitjano opened the pull request:
## Description

This PR mostly does some minor cleanups. It also improves the framegraph of the 3D axis: 

```
The axis framegraph has two distinct branches : one for the main
object (the axis or the cube) and one for the axis labels. Each of
those branches have a dedicated viewport. However, only one viewport
is needed since the two branches need the same part of the
window.

With this change, the first node of the axis' framegraph is now the
viewport (mViewport) and the two branches inherit from it:

- viewport
  <-- object branch -->
  - layerfilter (axisLayerFilter)
  - cameraselector
  - sortPolicy
  - clearBuffers

  <-- labels branch -->
  - layerfilter (twoDLayerFilter)
  - cameraselector
  - sortPolicy
  - clearBuffers

`from3DTo2DLabelPosition()` is also simplified to compute the
billboard position of the axis labels. Indeed, the viewport of the
labels had the dimensions of the main window (0, 0, 1, 1) while the
objects viewport has only the dimensions the relevant part of the
window. Therefore, `from3DTo2DLabelPosition()` needed to take into
this into account and recenter the computer coordinates. This is not
needed anymore. 
```

ptitjano committed:
qgs3daxis: Remove useless layer

The 3D Axis entities only need 2 layers:
- 1 to handle the visible object (the axis or the cube)
- 1 to handle the axis labels

The third one achieves no purpose. It can be safely removed.

ptitjano committed:
qgs3daxis: Give proper name to labels layer

ptitjano committed:
qgs3daxis: Simplify axis textchanged

`onTextXChanged`, `onTextYChanged` and `onTextZChanged` do the same
thing. Only the text entity (X, Y or Z) is different. With a simple
change, it is possible to define only one function.

ptitjano committed:
qgs3daxis: Give proper name to font parameters

ptitjano committed:
qgs3daxis: Use the same font everywhere

ptitjano committed:
qgs3daxis: Increase the axis size

This is achieved by decreasing the distance of the axis camera.

ptitjano committed:
qgs3daxis: Define only only one viewport in the framegraph

The axis framegraph has two distinct branches : one for the main
object (the axis or the cube) and one for the axis labels. Each of
those branches have a dedicated viewport. However, only one viewport
is needed since the two branches need the same part of the
window.

With this change, the first node of the axis' framegraph is now the
viewport (mViewport) and the two branches inherit from it:

- viewport
  <-- object branch -->
  - layerfilter (axisLayerFilter)
  - cameraselector
  - sortPolicy
  - clearBuffers

  <-- labels branch -->
  - layerfilter (twoDLayerFilter)
  - cameraselector
  - sortPolicy
  - clearBuffers

`from3DTo2DLabelPosition()` is also simplified to compute the
billboard position of the axis labels. Indeed, the viewport of the
labels had the dimensions of the main window (0, 0, 1, 1) while the
objects viewport has only the dimensions the relevant part of the
window. Therefore, `from3DTo2DLabelPosition()` needed to take into
this into account and recenter the computer coordinates. This is not
needed anymore.

ptitjano committed:
qgs3daxis: Fix some clang-tidy warnings

nyalldawson commented:
@ptitjano while you're working on this, valgrind (or asan) shows a bunch of leaks coming from the axis classes. Would it be possible to fix these too?

benoitdm-oslandia reviewed (approved):
Thanks for this rework!

ptitjano commented:
> @ptitjano while you're working on this, valgrind (or asan) shows a bunch of leaks coming from the axis classes. Would it be possible to fix these too?

Could you please share your output of the valgrind or explain or how you obtain it please ?

nyalldawson commented:
I don't have one on hand, but it's enough to just open a 3d view when running though valgrind. (From memory it also shows up if you run the tests under valgrind)

ptitjano commented:
> I don't have one on hand, but it's enough to just open a 3d view when running though valgrind. (From memory it also shows up if you run the tests under valgrind)

I don't see any leak when running the |test_3d_3drendering`. However, I do some information when running QGIS through valgrind. I get one error reported many times with one call. According to valgrind:

```
Qt3DExtras::QText2DEntity->setParent()
```
leaks. I don't understand how these calls could leak. All the qt3d entities entities in QGIS 3D are created with the same logic. Only the `QText2DEntity` are reported to leak. I think this is related to https://bugreports.qt.io/browse/QTBUG-77139

So, if this leaks, this is an upstream issue.

nyalldawson commented:
@ptitjano 

Here's the asan dump: https://gist.github.com/nyalldawson/802fa4d1e1c489eaef8f2482937e0271

nyalldawson commented:
@ptitjano 
I don't think it's related to https://bugreports.qt.io/browse/QTBUG-77139 -- it's not only the QText2DEntitys which are leaked, but everything created in Qgs3DAxis::createCube, and the mCubeRoot itself too. 

The issue is that in Qgs3DAxis::setEnableCube / Qgs3DAxis::setEnableAxis we are unsetting the parent for the disabled type, so those objects never get deleted (nor any of their children).

Qgs3DAxis::setEnableAxis / setEnableCube would need to be modified to delete the disabled object, instead of just unsetting the parent.

ptitjano committed:
qgs3daxis: Fix object root memory leak

If an object root entity (`mCubeRoot` or `mAxisRoot`) is not enabled,
it means that it does not have a parent because of the
`setEnableCube()` and `setEnableAxis()` logic. In that case, it will
never be automatically deleted by Qt. Therefore, it creates a memory
leak.

This issue is fixed by adding a check in the destructor. If an object
is not enabled, then it is disabled.

ptitjano commented:
> @ptitjano I don't think it's related to https://bugreports.qt.io/browse/QTBUG-77139 -- it's not only the QText2DEntitys which are leaked, but everything created in Qgs3DAxis::createCube, and the mCubeRoot itself too.
> 
> The issue is that in Qgs3DAxis::setEnableCube / Qgs3DAxis::setEnableAxis we are unsetting the parent for the disabled type, so those objects never get deleted (nor any of their children).
> 
> Qgs3DAxis::setEnableAxis / setEnableCube would need to be modified to delete the disabled object, instead of just unsetting the parent.

Thanks, the asan report is much complete. You're right, there is a memory leak. However, I'm afraid the solution is much more complex that just deleting the disabled object. The reason, we ended up with this solution is because if we delete the objects, we also need to delete the axis labels. Also, we need to properly update the raycaster and the associated hits entities. We tried to do so. The axis labels part is hard but it is possible to fix it without too much change. The raycaster part is much harder to fix. At some point during the interactions (from the menu or the configuration dialog) the raycaster logic stops working because the hits found logic does not work anymore. I tried to give it an other shot. I ended up completely rewriting the class. I definitely ended up introducing some bugs with this rewrite. 

Finally, I chose a much more conservative approach. The idea is to change the destructor to delete an object entity is it is deleted. This way, this fixes the memory leak and this does not introduce new issues.

nyalldawson commented on src/3d/qgs3daxis.cpp:
```diff
@@ -918,7 +898,7 @@ Qt3DExtras::QText2DEntity *Qgs3DAxis::addCubeText( const QString &text, float te
 }
 
-void Qgs3DAxis::createAxis( Qt::Axis axisType )
+void Qgs3DAxis::createAxis( const Qt::Axis &axisType )
```
```suggestion
void Qgs3DAxis::createAxis( Qt::Axis axisType )
```

It's actually preferable to pass small and trivially copiable types like enums by value, not reference.

ptitjano replied:
Dropped.

nyalldawson commented:
@ptitjano 

> Finally, I chose a much more conservative approach. The idea is to change the destructor to delete an object entity is it is deleted. This way, this fixes the memory leak and this does not introduce new issues.

Looks good to me! Thanks for this, it'll make tracking down other leaks in 3d views much easier (there was a LOT of noise before 😂)
